src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSProxyContext.cs

```diff
@@ -0,0 +1,609 @@
+        //  maintains a weak reference and notifies the managed side if the JS wrapper object
+        //  has been reclaimed by the JS GC. At that point, the managed side will release its
+        //  strong references, allowing the managed object to be collected.
+        // This ensures that things like delegates and promises will never 'go away' while JS
+        //  is expecting to be able to invoke or await them.
+        public IntPtr GetJSOwnedObjectGCHandle(object obj, GCHandleType handleType = GCHandleType.Normal)
+        {
+            if (obj == null)
+            {
+                return IntPtr.Zero;
+            }
+
+            lock (this)
+            {
+                if (ThreadJsOwnedObjects.TryGetValue(obj, out IntPtr gcHandle))
+                {
+                    return gcHandle;
+                }
+
+                IntPtr result = (IntPtr)GCHandle.Alloc(obj, handleType);
+                ThreadJsOwnedObjects[obj] = result;
+                return result;
+            }
+        }
+
+        // TODO unregister and collect pending PromiseHolder also when no C# is awaiting ?
+        public PromiseHolder GetPromiseHolder(nint gcHandle)
+        {
+            PromiseHolder holder;
+            if (IsGCVHandle(gcHandle))
+            {
+                lock (this)
+                {
+                    holder = new PromiseHolder(this, gcHandle);
+                    ThreadJsOwnedHolders.Add(gcHandle, holder);
+                }
+            }
+            else
+            {
+                holder = (PromiseHolder)((GCHandle)gcHandle).Target!;
+            }
+            return holder;
+        }
+
+        public unsafe void ReleaseJSOwnedObjectByGCHandle(nint gcHandle)
+        {
+            PromiseHolder? holder = null;
+            lock (this)
+            {
+                if (IsGCVHandle(gcHandle))
+                {
+                    if (ThreadJsOwnedHolders.Remove(gcHandle, out holder))
+                    {
+                        holder.GCHandle = IntPtr.Zero;
+                    }
+                    else
+                    {
+                        throw new InvalidOperationException("ReleaseJSOwnedObjectByGCHandle expected in ThreadJsOwnedHolders");
+                    }
+                }
+                else
+                {
+                    GCHandle handle = (GCHandle)gcHandle;
+                    var target = handle.Target!;
+                    if (target is PromiseHolder holder2)
+                    {
+                        holder = holder2;
+                        holder.GCHandle = IntPtr.Zero;
+                    }
+                    else
+                    {
+                        if (!ThreadJsOwnedObjects.Remove(target))
+                        {
+                            throw new InvalidOperationException("ReleaseJSOwnedObjectByGCHandle expected in ThreadJsOwnedObjects");
+                        }
+                    }
+                    handle.Free();
```

Comment:
can we guarantee by construction somehow that this handle will never get used again once we call `Free` on it? Looking at the current code it seems like the existing callers do cleanup but it would be cool if it was nearly impossible to use this incorrectly, since a gchandle double-free and a dangling gchandle can both have bad consequences. It's fine if there's no easy way to fix this though.

i.e. `void ReleaseJSOwnedObjectByGCHandle (ref nint gcHandle)` which zeroes out `gcHandle` before returning

Reply:
GCHandles are reused and I don't know any great way how to solve this.